### Como executar o código do zero

Para executar o código, apenas clique em 'Cell'(ou 'Runtime'), e depois 'Run All'

certifique-se de que os arquivos .csv estão no mesmo diretório do jupyter notebook

Link para download dos arquivos .csv:
    
[movies.csv](https://drive.google.com/file/d/1gLsCjaMrL91ECdThq58cZAzB9tPxG18g/view?usp=sharing)
    
[customers_rating.csv](https://drive.google.com/file/d/1C_T1w8fc7Oa8MeTo4LMTEcv90IfEOS-6/view?usp=sharing)

In [1]:
import pandas as pd

In [2]:
rdf = pd.read_csv("customers_rating.csv", sep=";") # Ratings DataFrame
rdf["Date"] = pd.to_datetime(rdf["Date"]) # Objetos do tipo Datetime do python

In [3]:
# Para o "movies.csv" dataset vamos alterar algumas coisas
mdf = pd.read_csv("movies.csv", sep=";", names=["Movie_Id", "Movies/Year"]) # Movies DataFrame
mdf = mdf.set_index("Movie_Id") # Id de cada filme será o index

mnames = [] # movie names == nome dos filmes
years = [] # Ano 

for item in mdf["Movies/Year"]: # Iteramos sobre a coluna 'Movies/Year'
    tmp = item.split(",") # Separamos cada elemento em uma lista com o primeiro elemento == filme e o segundo == Ano
    
    mnames.append(tmp[0].replace("(", "")) # 'replace' apenas para retirar os parentesis
    years.append(tmp[1].replace(")", ""))

mdf["Movies"] = mnames
mdf["Year"] = years

mdf.drop("Movies/Year", axis=1, inplace=True) # Retira a coluna agora inútil
mdf["Year"].astype('int') # Converte os anos para inteiros
print()

### Quantos filmes estão disponíveis no dataset?

Para isso iremos usar '.shape', que retorna as dimensões. Estamos interessados na quantidade de linhas, que representa a quantidade de filmes

In [4]:
mdf.shape[0]

4499

4499 filmes

### Qual o nome dos 5 filmes com melhor média de avaliação?

In [5]:
# 'groupby' nos permite realizar operações em um conjunto de dados, nesse caso na coluna com as notas(Rating) em que
# aplicamos a média com mean e depois sorteamos os valores em ordem decrescente
rdf.groupby(['Movie_Id']).Rating.mean().sort_values(ascending=False)[:5]

Movie_Id
3456    4.670989
3033    4.586364
2102    4.581296
4238    4.554434
13      4.552000
Name: Rating, dtype: float64

In [6]:
# loc para encontrar os filmes
print(mdf.loc[3456].Movies)
print(mdf.loc[3033].Movies)
print(mdf.loc[2102].Movies)
print(mdf.loc[4238].Movies)
print(mdf.loc[13].Movies)

Lost: Season 1
Ghost in the Shell: Stand Alone Complex: 2nd Gig
The Simpsons: Season 6
Inu-Yasha
Lord of the Rings: The Return of the King: Extended Edition: Bonus Material


### Quais os 5 anos com menos lançamentos de filmes?

In [7]:
mdf.Year.value_counts().sort_values(ascending=True)[:5]

 1926    1
 1915    1
 1922    1
 1917    1
 1929    2
Name: Year, dtype: int64

### Quantos filmes que possuem avaliação maior ou igual a 4.7, considerando apenas os filmes avaliados na última data de avaliação do dataset?

Quando foi a última data de avaliação do dataset?

In [8]:
rdf["Date"].sort_values(ascending=False)[:1]

10916395   2005-12-31
Name: Date, dtype: datetime64[ns]

In [9]:
date = "2005-12-31"
date = pd.to_datetime(date)
lrdf = rdf[rdf["Date"] == date] # Uma vez encontrada a data da ultima atualização, vamos separar uma variavel apenas
                                # com os valores que respeitam essa data

In [10]:
lrdf = lrdf.groupby(['Movie_Id']).Rating.mean().sort_values(ascending=False)

In [11]:
len(lrdf[lrdf>=4.7]) # Agora vemos quais os valores são maiores ou iguais a 4.7

195

### Quais os id's dos 5 customers que mais avaliaram filmes e quantas avaliações cada um fez?

In [12]:
rdf.Cust_Id.value_counts()[:5]

305344     4467
387418     4422
2439493    4195
1664010    4019
2118461    3769
Name: Cust_Id, dtype: int64

# FIM